endgame_simulations/endgame_simulation.py

```diff
@@ -0,0 +1,216 @@
+from abc import abstractproperty
+from typing import ClassVar, Generic, Iterator, Protocol, TypeVar, cast, overload
+
+import tqdm
+from hdf5_dataclass import FileType
+
+from endgame_simulations.models import (
+    BaseInitialParams,
+    BaseProgramParams,
+    EndgameModel,
+    _BaseUpdateParams,
+)
+
+from .common import AdvanceState, BaseState, State
+
+ProgramModel = TypeVar("ProgramModel", bound=BaseProgramParams)
+InitialParamsModel = TypeVar(
+    "InitialParamsModel", bound=BaseInitialParams, contravariant=True
+)
+UpdateParamsModel = TypeVar(
+    "UpdateParamsModel", bound=_BaseUpdateParams, contravariant=True
+)
+CombinedParams = TypeVar("CombinedParams", bound=BaseInitialParams, covariant=True)
+
+
+class CombineParams(
```

Comment:
I'm not sure I understand what is this class for. Can you please clarify? Is it intentional that its name is the same as the TypeVar defined above? It's a bit confusing what call actually returns.

Reply:
This is a mistake - revising

Reply:
Renamed in later version anyway